### PR TITLE
OCC commands to manage encryption

### LINKED ISSUE #15667: [enc2] API / OCC to enable encryption + select encryption module
To be able to automate setup for automated tests and also deployments, it is necessary to provide an API and/or CLI command to make it possible to enable the encryption module.

In the past, enabling encryption was only about enabling the "files_encryption" app.
Since 8.1 and encryption 2.0, enabling the "encryption" app is not enough. Admins need to go to the admin page and tick "enable encryption". An API/CLI command need to be provided to do this programmatically.

@schiesbn @th3fallen @DeepDiver1975 

CC @jnfrmarks

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #15667

@schiesbn including the two fixes we talked about:
1. Validate the module id before allowing to set it as default
2. Return the default module before returning the first item

@th3fallen @DeepDiver1975 @PVince81

### CODE REVIEW

**Comment by nickvergessen** [comment]
Will add tests tomorrow

**Comment by th3fallen** [comment]
> Will add tests tomorrow

@DeepDiver1975 's going to need a xanax when he reads this ;-)

**Comment by th3fallen** [comment]
Commands look good and work for me :+1:

**Comment by karlitschek** [comment]
:+1:

**Comment by nickvergessen** [comment]
Added tests for `setDefaultEncryptionModule()` and for returning the default module before the first item.
Then I deduplicated the Mock generation

**Comment by schiesbn** [comment]
we need to keep this PR https://github.com/owncloud/core/pull/15616 in mind which will change some stuff which is relevant for this PR. Maybe we should first finish #15616 

Keep this PR on hold for a moment... and I will discuss the next steps with @nickvergessen 

:-1: just that nobody merge it by accident.

**Comment by schiesbn** [comment]
https://github.com/owncloud/core/pull/15616 was merged so this pr can be rebased and adapted to the new return values.

**Comment by nickvergessen** [comment]
@schiesbn yes, was already on it :wink:

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please

File: lib/public/encryption/imanager.php [hunk 1]
```diff
@@ -78,13 +78,12 @@ function getEncryptionModules();
 	function getEncryptionModule($moduleId);
```
Comment (nickvergessen) [inline lib/public/encryption/imanager.php]: Missing public visibility, will fix as soon as jenkins reported back

**Comment by nickvergessen** [comment]
@owncloud-bot retest this please

**Comment by LukasReschke** [comment]
> :-1: just that nobody merge it by accident.

@schiesbn Time to re-review? :)

**Comment by carlaschroder** [comment]
You know what would be really super-awesome? Telling your poor old doc writer when you add new features :)

**Comment by pucela** [comment]
Yup


———————————
Carlos Montero-Luque
Vice President of Engineering
cml@owncloud.com

**Comment by nickvergessen** [comment]
@carlaschroder since the functionality is already present in the UI, this shouldn't be too much extra work. But sure. Will try to send a PR to the doc later, which can be merged once this here is merged.

**Comment by DeepDiver1975** [comment]
> Telling your poor old doc writer when you add new features :)

we should anyhow add new features to the wiki https://github.com/owncloud/core/wiki/ownCloud-8.1-Features

**Comment by carlaschroder** [comment]
> we should anyhow add new features to the wiki https://github.com/owncloud/core/wiki/
> ownCloud-8.1-Features

That works for me, as I refer to the wikis all the time.

**Comment by nickvergessen** [comment]
@owncloud-bot retest this please

**Comment by nickvergessen** [comment]
Documentation notes left at https://github.com/owncloud/documentation/issues/1053#issuecomment-95627319

**Comment by DeepDiver1975** [comment]
@schiesbn please review - THX

**Comment by schiesbn** [comment]
tested, works as expected... Code looks good :+1:

File: core/command/encryption/listmodules.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ListModules extends Base {
+	/** @var Manager */
+	protected $encryptionManager;
+
+	/**
+	 * @param Manager $encryptionManager
+	 */
+	public function __construct(Manager $encryptionManager) {
```
Comment (nickvergessen) [inline core/command/encryption/listmodules.php]: should use itnerface

**Comment by nickvergessen** [comment]
@owncloud-bot retest this please

**Comment by nickvergessen** [comment]
@owncloud-bot retest this please

**Comment by nickvergessen** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
```
./occ app:enable encryption
./occ app:enable encryption_dummy 
./occ encryption:enable
./occ encryption:list-modules
```

Leads to

```
{"reqId":"KjJnWqChd+DNqmogXYOA","remoteAddr":"","app":"PHP","message":"Argument 3 passed to OC\\Encryption\\Manager::registerEncryptionModule() must be callable, object given, called in \/home\/mjob\/Projekte\/owncloud\/master\/apps\/encryption_dummy\/appinfo\/app.php on line 5 and defined at \/home\/mjob\/Projekte\/owncloud\/master\/lib\/private\/encryption\/manager.php#98","level":0,"time":"2015-04-27T10:13:13+00:00","method":"--","url":"--"}
{"reqId":"KjJnWqChd+DNqmogXYOA","remoteAddr":"","app":"PHP","message":"Undefined index: default at \/home\/mjob\/Projekte\/owncloud\/master\/core\/command\/encryption\/listmodules.php#57","level":0,"time":"2015-04-27T10:13:13+00:00","method":"--","url":"--"}
{"reqId":"KjJnWqChd+DNqmogXYOA","remoteAddr":"","app":"PHP","message":"Undefined index: default at \/home\/mjob\/Projekte\/owncloud\/master\/core\/command\/encryption\/listmodules.php#57","level":0,"time":"2015-04-27T10:13:13+00:00","method":"--","url":"--"}
```

The first line is also present once dummy encryption is enabled.

**Comment by MorrisJobke** [comment]
Beside those warning everything works.

**Comment by nickvergessen** [comment]
the dummy module was fixed in another PR https://github.com/owncloud/core/pull/15718

**Comment by nickvergessen** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
Works now :+1:

**Pull request merged**